### PR TITLE
[DBZ-PGYB] Changes to support all operations with all replica identities

### PR DESCRIPTION
**Pull request by vaibhav-yb**
This PR adds the changes to the connector required to support all kinds of operations supported by YugabyteDB when coupled with various replica identities.

One change to note here would be that in case of `DEFUALT` replica identity, Postgres sends out the before image of the primary key column in case of `DELETE` and `UPDATE` operations. Meanwhile, YugabyteDB only sends out the before image of primary key in case of `DELETE` operations, and if a primary key itself is updated, YugabyteDB sends two events:
1. Delete for the existing row
2. Insert with the new value of the primary key

Additionally to test out everything, we have a new test class `YugabyteReplicaIdentityIT`.

### CODE REVIEW

File: debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/YugabyteReplicaIdentityIT.java [hunk 1]
```diff
@@ -0,0 +1,350 @@
...
+    // YB Note: Added a wait for replication slot to be active.
+    TestHelper.waitFor(Duration.ofSeconds(10));
+
+    waitForAvailableRecords(10_000, TimeUnit.MILLISECONDS);
+    // there shouldn't be any snapshot records
+    assertNoRecordsToConsume();
+
+    // insert and verify 2 new records
+    TestHelper.execute("INSERT INTO s2.a VALUES (1, 22, 'random text value');");
+    TestHelper.execute("UPDATE s2.a SET aa = 12345 WHERE pk = 1;");
```
Comment (Sumukh-Phalgaonkar) [inline debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/YugabyteReplicaIdentityIT.java:163]: Can you also add an Update where we set one of the columns to null. This will verify our logic for unchanged toasted datum for nulls.

File: debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/YugabyteReplicaIdentityIT.java [hunk 2]
```diff
@@ -0,0 +1,350 @@
...
+    try {
+      TestHelper.execute("DELETE FROM s2.a WHERE pk = 1;");
+    } catch (Exception sqle) {
+      assertThat(sqle.getMessage()).contains("ERROR: cannot delete from table \"a\" because it "
+                                              + "does not have a replica identity and publishes deletes");
+    }
+  }
+
+  @Test
+  public void shouldHaveBeforeImageForDeletesForReplicaIdentityFull() throws Exception {
```
Comment (Sumukh-Phalgaonkar) [inline debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/YugabyteReplicaIdentityIT.java:229]: Can we combine this with the test `shouldProduceOldValuesWithReplicaIdentityFull`. This will reduce the code duplication and number of tests in this file.

File: debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/YugabyteReplicaIdentityIT.java [hunk 3]
```diff
@@ -0,0 +1,350 @@
...
+    assertThat(deleteRecordValue.get(Envelope.FieldName.BEFORE)).isNotNull();
+
+    // Before field will have entries only for the primary key columns.
+    assertThat(deleteRecordValue.getStruct(Envelope.FieldName.BEFORE).getStruct("pk").getInt32("value")).isEqualTo(1);
+    assertThat(deleteRecordValue.getStruct(Envelope.FieldName.BEFORE).getStruct("aa").getInt32("value")).isNull();
+    assertThat(deleteRecordValue.getStruct(Envelope.FieldName.BEFORE).getStruct("bb").getString("value")).isNull();
+  }
+
+  @Test
+  public void shouldHaveBeforeImageForDeletesForReplicaIdentityChange() throws Exception {
```
Comment (Sumukh-Phalgaonkar) [inline debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/YugabyteReplicaIdentityIT.java:310]: Can we combine this with `shouldProduceEventsWithValuesForChangedColumnWithReplicaIdentityChange`.

File: debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/YugabyteReplicaIdentityIT.java [hunk 4]
```diff
@@ -0,0 +1,350 @@
...
+    assertThat(deleteRecordValue.get(Envelope.FieldName.BEFORE)).isNotNull();
+
+    // Before field will have entries for all the columns.
+    assertThat(deleteRecordValue.getStruct(Envelope.FieldName.BEFORE).getStruct("pk").getInt32("value")).isEqualTo(1);
+    assertThat(deleteRecordValue.getStruct(Envelope.FieldName.BEFORE).getStruct("aa").getInt32("value")).isEqualTo(22);
+    assertThat(deleteRecordValue.getStruct(Envelope.FieldName.BEFORE).getStruct("bb").getString("value")).isEqualTo("random text value");
+  }
+
+  @Test
+  public void shouldHaveBeforeImageForDeletesForReplicaIdentityDefault() throws Exception {
```
Comment (Sumukh-Phalgaonkar) [inline debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/YugabyteReplicaIdentityIT.java:269]: Can we combine this with `shouldProduceExpectedValuesWithReplicaIdentityDefault`

**Review by Sumukh-Phalgaonkar — CHANGES_REQUESTED** [review]

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresChangeRecordEmitter.java [hunk 5]
```diff
@@ -109,8 +109,9 @@ protected Object[] getOldColumnValues() {
                 case CREATE:
                     return null;
                 case UPDATE:
-                    // YB Note: For replica identity CHANGE, there is no old column value available.
-                    if (schema.getReplicaIdentity(tableId) == ReplicaIdentityInfo.ReplicaIdentity.CHANGE) {
+                    // YB Note: For replica identity CHANGE or DEFAULT, there is no old column value available.
+                    if (schema.getReplicaIdentity(tableId) == ReplicaIdentityInfo.ReplicaIdentity.CHANGE
+                            || schema.getReplicaIdentity(tableId) == ReplicaIdentityInfo.ReplicaIdentity.DEFAULT) {
```
[the remark this answers is not in the document]
Comment (vaibhav-yb) [reply debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresChangeRecordEmitter.java:114]: Even when the replica identity is `CHANGE` or `DEFAULT` the oldTupleList is not null, it just has the info about the columns, so the null check in the method is never hit and we need the check here.

File: debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/YugabyteReplicaIdentityIT.java (hunk 2 above)
Comment (vaibhav-yb) [reply debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/YugabyteReplicaIdentityIT.java:229]: The tests are written in JUnit4 and parameterising the tests there would require an extra effort of adding a separate runner class so I followed the test model Debezium is following in other test files and added duplicate test classes wherever required.

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresStreamingChangeEventSource.java [hunk 6]
```diff
@@ -175,29 +175,33 @@ public void execute(ChangeEventSourceContext context, PostgresPartition partitio
 
             this.lastCompletelyProcessedLsn = replicationStream.get().startLsn();
 
-            if (walPosition.searchingEnabled()) {
-                searchWalPosition(context, partition, this.effectiveOffset, stream, walPosition);
-                try {
-                    if (!isInPreSnapshotCatchUpStreaming(this.effectiveOffset)) {
-                        connection.commit();
+            // Against YB, filtering of records based on Wal position is only enabled when connector config provide.transaction.metadata is set to false.
+            if(!YugabyteDBServer.isEnabled() || (YugabyteDBServer.isEnabled() && !connectorConfig.shouldProvideTransactionMetadata())) {
```
Comment (suranjan) [inline debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresStreamingChangeEventSource.java:179]: space after if

**Review by suranjan — APPROVED** [review]

**Review by Sumukh-Phalgaonkar — APPROVED** [review]
